Database support

imenyoo2 opened the pull request:
i added a database support, so that the work time for each day get saved.
- the database has 2 column, days and worked_time

| days      | worked_time |
| ----------- | ----------- |
| 2022-08-11      | 4.013       |
| 2022-08-12   | 5.012        |

### how does it work?
- when the work session start running `START_TIME` get assigned to current time
- when the work session stops: current time mines `START_TIME` saved in the database

this is my first github contribution, any feedback would be appreciated

imenyoo2 committed:
feat: database support

unode commented on polypomo:
```diff
@@ -10,13 +10,18 @@ import time
 TOMATO = u"\U0001F345"  # Recommended font: Noto Emoji
 BREAK = u"\U0001F3D6"  # Recommended font: Noto Emoji
 PAUSE = u"\U000023F8"  # Recommended font: Siji
+DATABASE_SUPPORT = False  # change to Ture to save workedtime to database
```
Instead of a global to enable this feature, could you instead add a command-line argument that enables database logging?

There's also a small typo in this line: "Ture".

unode commented on polypomo:
```diff
@@ -133,6 +138,57 @@ class Status:
 
+@contextmanager
+def setup_conn():
+    global DATABASE_PATH
```
Not necessary given DATABASE_PATH isn't modified.

unode commented on polypomo:
```diff
@@ -133,6 +138,57 @@ class Status:
+            with setup_conn() as conn:
+                today = date.today()
+                worked_time = (time - START_TIME) / 60 / 60
+                todaysrow = conn.execute(f'SELECT * FROM data WHERE days = "{today}"').fetchall()
```
Can you use SQL arguments here instead of string replacing variables?
See https://stackoverflow.com/a/902417
Same for other SQL statements below

imenyoo2 committed:
feat:updating existing day rows

imenyoo2 committed:
final changes

unode commented on polypomo:
```diff
@@ -133,6 +138,57 @@ class Status:
+            conn.commit()
+            conn.close()
+
+def db_save(time, status):
```
time as an argument seems unnecessary here.
Can't you just execute `time.time()` inside the function?

unode commented on polypomo:
```diff
@@ -133,6 +138,57 @@ class Status:
+
+def db_save(time, status):
+    global START_TIME
+    if (status.status == 'work'):
```
Please move this if statement outside the function and format the style consistently (i.e. drop parenthesis)

unode commented:
Hi @imenyoo2,

Thanks a lot for this contribution.
I like the general idea but I would like to modify this PR into a more flexible logging solution.
To start with, I would prefer to have each individual sprint logged instead of daily summaries. The daily digest can always be generated from the data later. The functionality to aggregate and produce reports could live in a separate script that queries the database.

Apart from this I have several points of feedback that I will include as a review.

unode commented on polypomo:
```diff
@@ -10,13 +10,18 @@ import time
 BREAK = u"\U0001F3D6"  # Recommended font: Noto Emoji
 PAUSE = u"\U000023F8"  # Recommended font: Siji
+DATABASE_SUPPORT = False  # change to Ture to save workedtime to database
+DATABASE_PATH = os.getcwd() + '/polypomoDB.db'
```
Can this also be made into a command-line argument?

unode commented on polypomo:
```diff
@@ -133,6 +138,57 @@ class Status:
+    # check if the database exist
+    if (os.path.exists(DATABASE_PATH)):
+        print('database already exist')
+        conn = sqlite3.connect('polypomoDB.db')
```
What's the point of defining DATABASE_PATH if the file path is hardcoded here to "current directory"?

unode commented on polypomo:
```diff
@@ -133,6 +138,57 @@ class Status:
+        conn = sqlite3.connect('polypomoDB.db')
+        cur = conn.cursor()
+        cur.execute('''CREATE TABLE data
+               (days text, worked_time text)''')
```
This will need to be restructured following my comments in the pull request.
We'll want to store start/stop times in the database and compute a total day-time through external tools.

unode commented on polypomo:
```diff
@@ -133,6 +138,57 @@ class Status:
+def setup_conn():
+    global DATABASE_PATH
+    # check if the database exist
+    if (os.path.exists(DATABASE_PATH)):
```
This if clause can be removed if the SQL query below is converted to:
`CREATE TABLE IF NOT EXISTS`

unode commented on polypomo:
```diff
@@ -133,6 +138,57 @@ class Status:
+    global DATABASE_PATH
+    # check if the database exist
+    if (os.path.exists(DATABASE_PATH)):
+        print('database already exist')
```
Please consider removing these and other print statements below that read like debugging output

unode commented on polypomo:
```diff
@@ -133,6 +138,57 @@ class Status:
+        print('initiate database')
+        conn = sqlite3.connect('polypomoDB.db')
+        cur = conn.cursor()
+        cur.execute('''CREATE TABLE data
```
Most of the original code uses strings with (`"` - double quotes).
Could you use that throughout your code?

The present repository isn't configured to validate syntax and style but I'll seek to that after this.

unode commented on polypomo:
```diff
@@ -133,6 +138,33 @@ class Status:
+    if status == "work":
+        with setup_conn(path) as conn:
+            if  active:
+                conn.execute(f"INSERT INTO sessions VALUES (date('now'), time('now'), NULL)")
```
No need for f-string

unode commented on polypomo:
```diff
@@ -133,6 +138,33 @@ class Status:
+            if  active:
+                conn.execute(f"INSERT INTO sessions VALUES (date('now'), time('now'), NULL)")
+            else:
+                conn.execute(f"""UPDATE sessions
```
No need for f-string

unode commented on polypomo:
```diff
@@ -133,6 +138,33 @@ class Status:
+def db_save(status, path, active):
+    if status == "work":
+        with setup_conn(path) as conn:
+            if  active:
```
Can you remove the double space after `if` here

unode commented on polypomo:
```diff
@@ -133,6 +138,33 @@ class Status:
+        conn.commit()
+        conn.close()
+
+def db_save(status, path, active):
```
From the arguments this looks like it could be a member of `Status`.
I agree that conceptually its not the right class but it will simplify the function signature.

Can also be renamed to `save()` given we only have one way of saving at the moment.

unode commented on polypomo:
```diff
@@ -133,6 +138,33 @@ class Status:
+            else:
+                conn.execute(f"""UPDATE sessions
+                                 SET stop = time('now')
+                                 WHERE stop IS NULL
```
What if polypomo crashes and there's an incomplete entry from the past?
They will all have the same stop timestamp.
Can we maybe order by start time and `LIMIT 1` so this only affects the most recent record?

unode reviewed (changes requested):
This looks a lot cleaner than before. Thanks a lot for all the effort.
I think that if these comments are addressed I'd be happy to merge.

Please let me know as well which name I should credit this contribution under.

imenyoo2 commented:
Hi @unode 
thank you for you feedback, i just pushed a commit
here is what i did:
- i removed the global variables and replace them with a command line argument `--saveto`
- changed the name of the database table to `sessions`
- changed the structure of the database table to :

| date      | start | stop     |
| :---        |    :----:   |          ---: |
| 2022-08-11      | 18:43:31       | 18:43:33   |
| 2022-08-11   | 18:43:35        | 18:43:36      |

- i used sql functions `time()` and `date()` insted of python's

i didn't know what you meant by "moving the if statement outside of the function" thou, can you elaborate more?

imenyoo2 committed:
pool request feedback

imenyoo2 commented:
thanks a lot for you feedbacks, i enjoyed and learned a lot from this
experience
about the name, you can put my real name AYOUB AIT EL AOUAD

If there is anything else, please let me know.

On Thu, Aug 11, 2022 at 6:42 PM Renato Alves ***@***.***>
wrote:

> ***@***.**** requested changes on this pull request.
>
> This looks a lot cleaner than before. Thanks a lot for all the effort.
> I think that if these comments are addressed I'd be happy to merge.
>
> Please let me know as well which name I should credit this contribution
> under.
> ------------------------------
>
> In polypomo
> <https://github.com/unode/polypomo/pull/6#discussion_r943805256>:
>
> > +    cur.execute("""CREATE TABLE IF NOT EXISTS sessions (
> +                        date text,
> +                        start text,
> +                        stop text
> +        )""")
> +    try:
> +        yield cur
> +    finally:
> +        conn.commit()
> +        conn.close()
> +
> +def db_save(status, path, active):
> +    if status == "work":
> +        with setup_conn(path) as conn:
> +            if  active:
> +                conn.execute(f"INSERT INTO sessions VALUES (date('now'), time('now'), NULL)")
>
> No need for f-string
> ------------------------------
>
> In polypomo
> <https://github.com/unode/polypomo/pull/6#discussion_r943805343>:
>
> > +                        start text,
> +                        stop text
> +        )""")
> +    try:
> +        yield cur
> +    finally:
> +        conn.commit()
> +        conn.close()
> +
> +def db_save(status, path, active):
> +    if status == "work":
> +        with setup_conn(path) as conn:
> +            if  active:
> +                conn.execute(f"INSERT INTO sessions VALUES (date('now'), time('now'), NULL)")
> +            else:
> +                conn.execute(f"""UPDATE sessions
>
> No need for f-string
> ------------------------------
>
> In polypomo
> <https://github.com/unode/polypomo/pull/6#discussion_r943808939>:
>
> > +    cur = conn.cursor()
> +    cur.execute("""CREATE TABLE IF NOT EXISTS sessions (
> +                        date text,
> +                        start text,
> +                        stop text
> +        )""")
> +    try:
> +        yield cur
> +    finally:
> +        conn.commit()
> +        conn.close()
> +
> +def db_save(status, path, active):
> +    if status == "work":
> +        with setup_conn(path) as conn:
> +            if  active:
>
> Can you remove the double space after if here
> ------------------------------
>
> In polypomo
> <https://github.com/unode/polypomo/pull/6#discussion_r943810739>:
>
> > +def setup_conn(path):
> +    # check if the database exist
> +    conn = sqlite3.connect(path)
> +    cur = conn.cursor()
> +    cur.execute("""CREATE TABLE IF NOT EXISTS sessions (
> +                        date text,
> +                        start text,
> +                        stop text
> +        )""")
> +    try:
> +        yield cur
> +    finally:
> +        conn.commit()
> +        conn.close()
> +
> +def db_save(status, path, active):
>
> From the arguments this looks like it could be a member of Status.
> I agree that conceptually its not the right class but it will simplify the
> function signature.
>
> Can also be renamed to save() given we only have one way of saving at the
> moment.
> ------------------------------
>
> In polypomo
> <https://github.com/unode/polypomo/pull/6#discussion_r943813110>:
>
> > +        )""")
> +    try:
> +        yield cur
> +    finally:
> +        conn.commit()
> +        conn.close()
> +
> +def db_save(status, path, active):
> +    if status == "work":
> +        with setup_conn(path) as conn:
> +            if  active:
> +                conn.execute(f"INSERT INTO sessions VALUES (date('now'), time('now'), NULL)")
> +            else:
> +                conn.execute(f"""UPDATE sessions
> +                                 SET stop = time('now')
> +                                 WHERE stop IS NULL
>
> What if polypomo crashes and there's an incomplete entry from the past?
> They will all have the same stop timestamp. Can we maybe order by start
> time and LIMIT 1 so this only affects the most recent record?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/unode/polypomo/pull/6#pullrequestreview-1070176693>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AUZIEJTHMOXLPDWBWLPCAYLVYVCRHANCNFSM56IACN4Q>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

imenyoo2 committed:
moved db_save to State and modified updating SQL command

imenyoo2 committed:
final changes

unode commented on polypomo:
```diff
@@ -132,6 +137,37 @@ class Status:
+                    conn.execute("INSERT INTO sessions VALUES (date('now'), time('now'), NULL)")
+                else:
+                    conn.execute("""UPDATE sessions
+                                     SET stop = time('now')
+                                     WHERE start IN (
+                                         SELECT start FROM sessions
+                                         ORDER BY start DESC
+                                         LIMIT 1)
```
```suggestion
                                     SET stop = time('now')
                                     WHERE stop is NULL
                                     ORDER BY start DESC
                                     LIMIT 1)
```

Can you check if this alternative query works?

imenyoo2 replied:
this query doesn't work
```
Parse error: near "order": syntax error
  essions set stop = time('now') where stop is null order by start desc limit 1;
                                      error here ---^
```

unode replied:
Seems like it's a limitation of SQLite - https://stackoverflow.com/a/29074519
Thanks for trying it out.

unode commented on polypomo:
```diff
@@ -103,6 +106,8 @@ class Status:
 
     def toggle(self):
         self.active = not self.active
+        if self.saveto:
```
```suggestion
        if self.saveto and self.status == "work":
```

unode commented on polypomo:
```diff
@@ -132,6 +137,37 @@ class Status:
             self.timer = Timer(self.worktime)
 
+    def save(self):
+        if self.status == "work":
```
If you accept the suggestion above, this `if` can now be removed from `save()`.

unode reviewed (commented):
This is the `if` I was referring to before.

unode commented:
Can you also add a couple of sentences to the README to document the new command-line parameter and how to use it from polypomo?

imenyoo2 committed:
Update polypomo

Co-authored-by: Renato Alves <alves.rjc@gmail.com>

imenyoo2 committed:
update README

unode commented:
Thanks a lot for proposing, implementing it and taking into account all the feedback and improving it beyond my suggestions.

I'm quite happy with the final result. Really nice work!
